src/python/pants/backend/url_handlers/s3/register.py

```diff
@@ -69,36 +70,47 @@ async def download_from_s3(
 ) -> Digest:
     from botocore import auth, compat, exceptions  # pants: no-infer-dep
 
-    # NB: The URL for auth is expected to be in path-style
-    path_style_url = "https://s3"
+    virtual_hosted_url = f"https://{request.bucket}.s3.amazonaws.com/{request.key}"
     if request.region:
-        path_style_url += f".{request.region}"
-    path_style_url += f".amazonaws.com/{request.bucket}/{request.key}"
+        virtual_hosted_url = (
+            f"https://{request.bucket}.s3.{request.region}.amazonaws.com/{request.key}"
+        )
+
     if request.query:
-        path_style_url += f"?{request.query}"
+        virtual_hosted_url += f"?{request.query}"
 
     headers = compat.HTTPHeaders()
     http_request = SimpleNamespace(
-        url=path_style_url,
+        url=virtual_hosted_url,
         headers=headers,
         method="GET",
         auth_path=None,
+        data=None,
+        params={},
+        context={},
+        body={},
     )
+
     # NB: The added Auth header doesn't need to be valid when accessing a public bucket. When
     # hand-testing, you MUST test against a private bucket to ensure it works for private buckets too.
-    signer = auth.HmacV1Auth(aws_credentials.creds)
+    # adding x-amz-content-SHA256 as per boto code
+    # ref link - https://github.com/boto/botocore/blob/547b20801770c8ea4255ee9c3b809fea6b9f6bc4/botocore/auth.py#L52C1-L54C2
+    headers.add_header(
+        "X-Amz-Content-SHA256",
+        EMPTY_FILE_DIGEST.fingerprint,
+    )
+
+    # We require a region to sign the request with sigv4
+    # If we don't know where the bucket is, default to the region from the credentials
+    # and fallback to us-east-1
+    signing_region = request.region or aws_credentials.default_region or "us-east-1"
+
+    signer = auth.SigV4Auth(aws_credentials.creds, "s3", signing_region)
```

Comment:
Is it worth supporting the old codepath under a flag (HmacV1Auth)? Not sure risky you view this change as

Reply:
Good idea. If on the remote chance a user does see an issue, they can just configure the old behavior. (S3 might be perfectly fine with it, but I can imagine a user using some S3 API-compatible service which we have never heard of and having an issue. May never happen but I can't discount the possibility. Feature flags are cheap insurance.

You can set a `removal_version` and `removal_hint` on the transition option so that we maintainers know to remove the option at an appropriate point in the future (or reevaluate its necessity at least,  maybe document that in the `removal_hint`).

Reply:
Yeah I don't see one either. Maybe add a new one then?

Reply:
Maybe `aws-s3-download-handler` or a better name?

Reply:
@tdyas I've added the hmacv1 back along with a new option subsystem to enable it (defaulting to sigv4) I wasn't sure deprecation is really necessary so I omitted it for now. If you'd like me to add it I can though, how many versions out should I target?